rust/foxglove/src/websocket.rs

```diff
@@ -143,31 +141,145 @@ pub(crate) struct ConnectedClient {
     weak_self: Weak<Self>,
     sink_id: SinkId,
     context: Weak<Context>,
+    poller: parking_lot::Mutex<Option<Poller>>,
     /// A cache of channels for `on_subscribe` and `on_unsubscribe` callbacks.
     channels: parking_lot::RwLock<HashMap<ChannelId, Arc<RawChannel>>>,
-    /// Write side of a WS stream
-    sender: Mutex<WebsocketSender>,
     data_plane_tx: flume::Sender<Message>,
     data_plane_rx: flume::Receiver<Message>,
     control_plane_tx: flume::Sender<Message>,
-    control_plane_rx: flume::Receiver<Message>,
     service_call_sem: Semaphore,
     fetch_asset_sem: Semaphore,
     /// Subscriptions from this client
     subscriptions: parking_lot::Mutex<BiHashMap<ChannelId, SubscriptionId>>,
     /// Channels advertised by this client
     advertised_channels: parking_lot::Mutex<HashMap<ClientChannelId, Arc<ClientChannel>>>,
     server: Weak<Server>,
-    /// The cancellation_token is used by the server to disconnect the client.
-    /// It's cancelled when the client's control plane queue fills up (slow client).
-    cancellation_token: CancellationToken,
+    shutdown_tx: parking_lot::Mutex<Option<oneshot::Sender<ShutdownReason>>>,
+}
+
+/// A poller for a connected client.
+///
+/// The poller is repsonsible for:
+/// - Sending messages (from `data_plane` and `control_plane`) to the websocket.
+/// - Receiving messages from the websocket and invoking [`ConnectedClient::handle_message`].
+/// - Waiting for a shutdown signal, and closing the websocket.
+struct Poller {
+    websocket: WebSocketStream<TcpStream>,
+    data_plane_rx: flume::Receiver<Message>,
+    control_plane_rx: flume::Receiver<Message>,
+    shutdown_rx: oneshot::Receiver<ShutdownReason>,
+}
+
+enum ShutdownReason {
```

Comment:
This is a nice way of handling shutdown for different reasons, it solves a lot of problems very neatly!